### DISCUSSION TITLE
Run function on window startup without stopping the UI

### DISCUSSION THREAD

**Discussion by xavizardKnight** [AvaloniaUI/Avalonia#6520] [Q&A] [answered]
Hello! I'm writing a tool for making backups of the savedata of a game. I have a window which sole propose is to run the backup while the window shows the progress of the backup (how many files are copied, how many are left, percentage of the copy, etc…)

The problem is, no matter how I do it, the entire backup gets completed before the window appears.
The function `DoTheThing()` (I'm bad at naming things) is the function that starts the backup.

I tried placing it inside the `BackupProgressWindow()`, inside `InitializeComponent()` and as methods of the window `GotFocus`, `OnActivated` or `OnInitialized` events, all with the same result; backup completes first then window appears.

Here's how my BackupProgressWindow.axaml.cs looks currently:

```cs
		public BackupProgressWindow()
		{
			InitializeComponent();
#if DEBUG
			this.AttachDevTools();
#endif
			DoTheThing();
		}

		private void InitializeComponent()
		{
			AvaloniaXamlLoader.Load(this);
		}
		
		public void DoTheThing()
		{
			timeStart = DateTime.Now;
			CommonSteps();
			switch(type)
			{
				case 1: //From game to folder, directly
					SimpleBackup();
					break;
			}
		}
```

How can I make it so the `DoTheThing()` function (and all the functions that are called inside of DoTheThing) are executed while the window is refreshed?

**Comment by sn4k3** [comment, marked as answer]
If that is the only objective, why not do a .bat/bash/sh script? Or a console app?
Anyway to do what you want you need a background task that copies the files and report back to the UI via the Dispatcher

See: https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/task-based-asynchronous-programming

While on your task:
```C#
Dispatcher.UIThread.InvokeAsync(() => ReportMyProgress(copiedFiles));
```

**Reply by xavizardKnight** [reply]
Thank you! This line is the one I needed!

**Comment by maxkatz6** [comment]
https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/
